main.go

```diff
@@ -5,13 +5,27 @@ import (
 	"net/http"
 )
 
+type State struct {
+	MyHasher
+	MyKVS
+	Stats
+}
+
 func main() {
-	hasher := NewMyHasher()
-	kv := NewMyKVS()
+	state := &State{NewMyHasher(),
+		NewMyKVS(),
+		Stats{
+			ServerStats{},
+			make([]Handler, 3),
+		},
+	}
+	state.Stats.Handlers[Redirect] = Handler{"/", 0}
+	state.Stats.Handlers[Shorten] = Handler{"/shorten", 0}
+	state.Stats.Handlers[Statistics] = Handler{"/statistics", 0}
 
-	http.HandleFunc("/", root(kv))
-	http.HandleFunc("/shorten/", shorten(kv, hasher))
-	http.HandleFunc("/stats", stats())
+	http.HandleFunc("/", redirect(state))
+	http.HandleFunc("/shorten/", shorten(state))
+	http.HandleFunc("/statistics", statistics(state))
```

Comment:
Since the 3 functions here (`redirect` `shorten` and `statistics` take a State struct) it would probably seem a bit more natural if all 3 were methods of the `State` struct. I think the code would gain in clarity, though it may be subjective

To illustrate my point here how that would look if redirect was a method of State (which I renamed to `Server` since at the end of the day this struct represents our url-shortening server):

```go
server := Server{...}

http.HandleFunc("/", server.redirect)

func (s*Server) redirect (w http.ResponseWriter, r *http.Request) {
    // no need to return a closure any more here
    updateStats(state, Redirect)

    // ... rest of the code
}
```